edb/server/tenant.py

```diff
@@ -1563,6 +1640,92 @@ def set_readiness_state(self, state: srvargs.ReadinessState, reason: str):
+            await self._reload_config_file()
+        except Exception:
+            logger.error("failed to reload config file", exc_info=True)
+            metrics.background_errors.inc(
+                1.0, self._instance_name, "reload_config_file"
+            )
+
+    async def load_config_file(self, compiler):
+        logger.info("loading config file")
+
+        # Read the TOML file
+        with self._config_file.open('rb') as f:
+            toml_data = tomllib.load(f)
+
+        # Handle special case for `magic_smtp_config`
+        magic_smtp_config = toml_data.pop("magic_smtp_config", None)
+        if magic_smtp_config:
+            await self.load_sidechannel_configs(
+                magic_smtp_config, compiler=compiler
+            )
+
+        # Parse TOML config file content into JSON
+        if toml_data and toml_data.get("cfg::Config"):
+            result = compiler.compile_structured_config(
+                toml_data, "configuration file"
+            )
+            if asyncio.iscoroutine(result):
+                result = await result
+
+            def setting_filter(value: config.SettingValue) -> bool:
+                if self._server.config_settings[value.name].backend_setting:
+                    raise errors.ConfigurationError(
+                        f"backend config {value.name!r} cannot be set "
+                        f"via config file"
+                    )
+                return True
+
+            json_obj = config.to_json_obj(
+                self._server.config_settings,
+                result["cfg::Config"],
+                include_source=False,
+                setting_filter=setting_filter,
+            )
+            config_file_data = [
+                {
+                    "name": name,
+                    "value": value,
+                    "type": config.ConStateType.config_file,
+                }
+                for name, value in json_obj.items()
+            ]
+        else:
+            config_file_data = []
+
+        # Update init_con_data and SQL
+        self.set_init_con_data(
+            [
+                cs
+                for cs in self._init_con_data
+                if cs["type"] != config.ConStateType.config_file
+            ]
+            + config_file_data
+        )
+
+    async def _reload_config_file(self):
+        # Load TOML config file
+        compiler = self._server.get_compiler_pool()
+        await self.load_config_file(compiler)
+
+        # Update sys pgcon and reload system config
+        async with self.use_sys_pgcon() as syscon:
+            if syscon.last_init_con_data is not self._init_con_data:
+                await syscon.sql_execute(
+                    pgcon.RESET_STATIC_CFG_SCRIPT + (self._init_con_sql or b'')
+                )
+                syscon.last_init_con_data = self._init_con_data
+            sys_config = await self._load_sys_config(syscon=syscon)
+            # GOTCHA: don't notify the change of sysconfig because it's local
```

Comment:
Hm?

Reply:
Ah, don't notify other tenants on the same backend; comment updated.